google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/Job.java

```diff
@@ -238,12 +239,23 @@ public boolean isDone() {
    *     to complete
    */
   public Job waitFor(RetryOption... waitOptions) throws InterruptedException {
+    return waitForInternal(DEFAULT_RETRY_CONFIG, waitOptions);
+  }
+
+  public Job waitFor(BigQueryRetryConfig bigQueryRetryConfig, RetryOption... waitOptions)
+      throws InterruptedException {
+    return waitForInternal(bigQueryRetryConfig, waitOptions);
+  }
+
+  private Job waitForInternal(BigQueryRetryConfig bigQueryRetryConfig, RetryOption... waitOptions)
+      throws InterruptedException {
     checkNotDryRun("waitFor");
     Object completedJobResponse;
     if (getConfiguration().getType() == Type.QUERY) {
       completedJobResponse =
           waitForQueryResults(
               RetryOption.mergeToSettings(DEFAULT_JOB_WAIT_SETTINGS, waitOptions),
+              bigQueryRetryConfig,
               DEFAULT_QUERY_WAIT_OPTIONS);
     } else {
       completedJobResponse =
```

Comment:
Does this change need to plumb the new retry config down through the non-query path (`waitForJob`) as well?

Reply:
The current behavior for non-query path does not retry on BigQuery error messages so it does not need to be plumb through. I think we can consider adding retry on BigQuery error messages for non-query path also but it would change the existing default behavior so lets discuss offline and add it on a later PR if needed.

I've updated the doc string to make it more clear. PTAL.